lib/mongo/protocol.ex

```diff
@@ -125,7 +131,17 @@ defmodule Mongo.Protocol do
   end
 
   def handle_execute(%Mongo.Query{action: action, extra: extra}, params, opts, s) do
-    handle_execute(action, extra, params, opts, s)
+    :ok = :inet.setopts(s.socket, [active: false])
+    with {:ok, reply, new_s} <- handle_execute(action, extra, params, opts, %{
+           s | database: Keyword.get(opts, :database, s.database)
+         }) do
```

Comment:
Yes, we need the separation. This is allowing you to specify a different database for the specified operation, which updates the state's database temporarily. Since `handle_execute` returns a modified state, and we handed it the state with the temporary database, we need the old state to restore the original database. I've renamed the variables to make it clearer.